src/vs/platform/terminal/electron-main/electronPtyHostStarter.ts

```diff
@@ -4,60 +4,99 @@
 	constructor(
 		private readonly _reconnectConstants: IReconnectConstants,
-		@IEnvironmentService private readonly _environmentService: INativeEnvironmentService
+		@IEnvironmentService private readonly _environmentService: INativeEnvironmentService,
+		@ILifecycleMainService private readonly _lifecycleMainService: ILifecycleMainService,
+		@ILogService private readonly _logService: ILogService
 	) {
 	}
 
 	start(lastPtyId: number): IPtyHostConnection {
-		return new NodePtyHostStarter(this._reconnectConstants, this._environmentService).start(lastPtyId);
-
-		// console.log('use utility proc');
-
-		// // TODO: Convert to use utility process
-		// const opts: IIPCOptions = {
-		// 	serverName: 'Pty Host',
-		// 	args: ['--type=ptyHost', '--logsPath', this._environmentService.logsHome.fsPath],
-		// 	env: {
-		// 		VSCODE_LAST_PTY_ID: lastPtyId,
-		// 		VSCODE_PTY_REMOTE: this._isRemote,
-		// 		VSCODE_AMD_ENTRYPOINT: 'vs/platform/terminal/node/ptyHostMain',
-		// 		VSCODE_PIPE_LOGGING: 'true',
-		// 		VSCODE_VERBOSE_LOGGING: 'true', // transmit console logs from server to client,
-		// 		VSCODE_RECONNECT_GRACE_TIME: this._reconnectConstants.graceTime,
-		// 		VSCODE_RECONNECT_SHORT_GRACE_TIME: this._reconnectConstants.shortGraceTime,
-		// 		VSCODE_RECONNECT_SCROLLBACK: this._reconnectConstants.scrollback
-		// 	}
-		// };
-
-		// const ptyHostDebug = parsePtyHostDebugPort(this._environmentService.args, this._environmentService.isBuilt);
-		// if (ptyHostDebug) {
-		// 	if (ptyHostDebug.break && ptyHostDebug.port) {
-		// 		opts.debugBrk = ptyHostDebug.port;
-		// 	} else if (!ptyHostDebug.break && ptyHostDebug.port) {
-		// 		opts.debug = ptyHostDebug.port;
-		// 	}
-		// }
-
-		// const client = new Client(FileAccess.asFileUri('bootstrap-fork').fsPath, opts);
-
-		// return {
-		// 	client,
-		// 	dispose: client.dispose,
-		// 	onDidProcessExit: client.onDidProcessExit
-		// };
+		this.utilityProcess = new UtilityProcess(this._logService, NullTelemetryService, this._lifecycleMainService);
+
+		const inspectParams = parsePtyHostDebugPort(this._environmentService.args, this._environmentService.isBuilt);
+		let execArgv: string[] | undefined = undefined;
+		if (inspectParams) {
+			execArgv = ['--nolazy'];
+			if (inspectParams.break) {
+				execArgv.push(`--inspect-brk=${inspectParams.port}`);
+			} else if (!inspectParams.break) {
+				execArgv.push(`--inspect=${inspectParams.port}`);
+			}
+		}
+
+		this.utilityProcess.start({
+			type: 'ptyHost',
+			entryPoint: 'vs/platform/terminal/node/ptyHostMain',
+			payload: this._createPtyHostConfiguration(lastPtyId),
+			execArgv
+		});
+
+		const port = this.utilityProcess.connect();
+		const client = new MessagePortClient(port, 'ptyHost');
+
+		// Listen for new windows to establish connection directly to pty host
+		validatedIpcMain.on('vscode:createPtyHostMessageChannel', (e, nonce) => this._onWindowConnection(e, nonce));
+
+		// TODO: Do we need to listen for window close to close the port?
```

Comment:
Do we need to clean up the port?

Reply:
Well, maybe @deepak1556 could clarify if a `MessagePort` that is inside a renderer window is automatically closing when the window goes down.

For the file watcher, I have this code to explicitly kill a watcher utility process when the window either changes to another workspace (`LOAD` event) or closes:

https://github.com/microsoft/vscode/blob/c7581b570fcba2dff2a026495eb41c21e354aeac/src/vs/platform/utilityProcess/electron-main/utilityProcess.ts#L469-L478

In your case this is maybe not needed because the PTY host continues to exist even after the window lifecycle ends.

Reply:
> Well, maybe @deepak1556 could clarify if a MessagePort that is inside a renderer window is automatically closing when the window goes down.

Yes and you will receive a `close` event on the other end of the port you are listening on.